components/server/src/messaging/redis-subscriber.ts

```diff
@@ -0,0 +1,120 @@
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+import { RedisClient } from "../redis/client";
+import {
+    HeadlessWorkspaceEventListener,
+    LocalMessageBroker,
+    PrebuildUpdateListener,
+    WorkspaceInstanceUpdateListener,
+} from "./local-message-broker";
+import { inject, injectable } from "inversify";
+import {
+    Disposable,
+    DisposableCollection,
+    RedisWorkspaceInstanceUpdate,
+    WorkspaceInstanceUpdatesChannel,
+} from "@gitpod/gitpod-protocol";
+import { log } from "@gitpod/gitpod-protocol/lib/util/logging";
+import { getExperimentsClientForBackend } from "@gitpod/gitpod-protocol/lib/experiments/configcat-server";
+import { Attributes } from "@gitpod/gitpod-protocol/lib/experiments/types";
+import { reportRedisUpdateCompleted, reportRedisUpdateReceived } from "../prometheus-metrics";
+
+@injectable()
+export class RedisSubscriber implements LocalMessageBroker {
+    constructor(@inject(RedisClient) private readonly redis: RedisClient) {}
+
+    protected workspaceInstanceUpdateListeners: Map<string, WorkspaceInstanceUpdateListener[]> = new Map();
+
+    protected readonly disposables = new DisposableCollection();
+
+    async start(): Promise<void> {
+        const channels = [WorkspaceInstanceUpdatesChannel];
+        const client = this.redis.get();
+
+        for (let chan of channels) {
+            await client.subscribe(chan);
+            this.disposables.push(Disposable.create(() => client.unsubscribe(chan)));
+        }
+
+        client.on("message", async (channel: string, message: string) => {
+            reportRedisUpdateReceived(channel);
+
+            const featureEnabled = await this.isRedisPubSubEnabled({});
+            if (!featureEnabled) {
+                log.debug("[redis] Redis listener is disabled through feature flag", { channel, message });
+                return;
+            }
+
+            let err: Error | undefined;
+            try {
+                await this.onMessage(channel, message);
+                log.debug("[redis] Succesfully handled update", { channel, message });
+            } catch (e) {
+                err = e;
+                log.error("[redis] Failed to handle message from Pub/Sub", { channel, message });
+            } finally {
+                reportRedisUpdateCompleted(channel, err);
+            }
+        });
+    }
+
+    private async onMessage(channel: string, message: string): Promise<void> {
+        switch (channel) {
+            case "chan:instances":
+                const parsed = JSON.parse(message) as RedisWorkspaceInstanceUpdate;
+                return this.onInstanceUpdate(parsed);
+            default:
+                throw new Error(`Redis Pub/Sub received message on unknown channel: ${channel}`);
+        }
+    }
+
+    private async onInstanceUpdate(update: RedisWorkspaceInstanceUpdate): Promise<void> {
+        log.debug("[redis] Received instance update", { update });
```

Comment:
For now, we just log to validate the pipeline is working. We've also not got any subscribers at this point (we've not added those invocation hooks yet)